composable_kernel/include/tensor_operation/element_wise_operation.hpp

```diff
@@ -100,6 +100,105 @@ struct AddHardswishAdd
     }
 };
 
+// Unary operators are usually called element-wisely before the reduction is executed on the
+// elements.
+// They are needed for easy implementation of reduction types of AVG, NRM1, NRM2
+
+template <typename Y, typename X, bool hasDividing = false> 
+struct unary_identic
+{
+    __host__ __device__ unary_identic(const int divider = 1) { (void)divider; };
+
+    __host__ __device__ inline constexpr void operator()(Y& y, const X& x) const
+    {
+        y = type_convert<Y>(x);
+    };
+};
+
+template <typename Y, typename X>
+struct unary_identic<Y, X, true>
+{   
+    __host__ __device__ unary_identic(const int divider = 1)
+    {
+        scaler = 1.0f / static_cast<float>(divider);
+    };
+    
+    __host__ __device__ inline constexpr void operator()(Y& y, const X& x) const
+    {
+        y = type_convert<Y>(x) * type_convert<Y>(scaler);
+    };
+
+    float scaler = 1.0f;
+};
+
+template <typename Y, typename X, bool hasDividing = false>
+struct unary_square
+{
+    __host__ __device__ unary_square(const int divider = 1) { (void)divider; };
+
+    __host__ __device__ inline constexpr void operator()(Y& y, const X& x) const
+    {
+        y = type_convert<Y>(x) * type_convert<Y>(x);
+    };
+};
+
+template <typename Y, typename X>
+struct unary_square<Y, X, true>
+{
+    __host__ __device__ unary_square(const int divider = 1)
+    {
+        scaler = 1.0f / static_cast<float>(divider);
+    };
+
+    __host__ __device__ inline constexpr void operator()(Y& y, const X& x) const
+    {
+        y = type_convert<Y>(x) * type_convert<Y>(x) * type_convert<Y>(scaler);
+    };
+
+    float scaler = 1.0f;
+};
+
+static inline __device__ half_t abs(half_t x) { return __habs(x); };
+static inline __device__ half_t sqrtf(half_t x) { return hsqrt(x); };
+
+template <typename Y, typename X, bool hasDividing = false>
+struct unary_abs
+{
+    __host__ __device__ unary_abs(const int divider = 1) { (void)divider; };
+
+    __host__ __device__ inline constexpr void operator()(Y& y, const X& x) const
```

Comment:
> please sqrtf and abs in math.hpp

Here `sqrtf` and `abs` are instantiation for half_t, only used by my Unary operators.  Are you sure to move them to `math.hpp` ?  If moved to `math.hpp`, then `math.hpp` need include `data_types.hpp`,   but actually don't know which file is more fundamental.

Reply:
yes, please implement ```sqrt``` and ```abs``` in math.hpp. It's ok to include data_type.hpp in math.hpp

Reply:
I found let math.hpp to include data_type.hpp will cause  dead-lock including, since already have `data_type.hpp` include `statically_indexed_array.hpp` include `sequence.hpp` include `math.hpp`

Reply:
Then you can move sqrt and abs into math_v2.hpp